Compilation error

**Describe the bug**

I got my first situation now where the initial syntax check or so that stan is doing passes, but the compilation fails. And I don’t really understand why or how I can start to debug this - I tried to change the code for an hour and it still fails. I cannot reproduce the compilation error when having the single function in a separate stan file.

**To Reproduce**

Apologies - I could not reproduce this bug so far in a minimum example. 
The problem comes down to a block of code like this:

```
matrix fun(matrix time) {
    matrix[rows(time), cols(time)] result;
    real val;
    val = time[1, 2];
    return result;
}
```

(note that this is simplified but it fails like this) It starts compiling but after few seconds gives a huge error message, that ends with:

```
stan/src/stan/model/indexing/rvalue.hpp:55:10: note:   template argument deduction/substitution failed:
/tmp/RtmpiGD5kC/model-4fe3695a40a9.hpp:436:17: note:   candidate expects 2 arguments, 4 provided
  436 |     val = rvalue(time, "time", index_uni(1), index_uni(2));
      |           ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
make: *** [make/program:58: /tmp/RtmpiGD5kC/model-4fe3695a40a9] Fehler 1
Fehler: An error occured during compilation! See the message above for more information.
```

**Expected behavior**

I would have expected that the code compiles without error.

**Operating system**

Ubuntu 20.04.3 LTS

**CmdStanR version number**

CmdStan 2.28.2 via cmdstanr_0.4.0

**Additional context**

Compiler/Toolkit: g++ (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0

--- a/Jenkinsfile
+++ b/Jenkinsfile
@@ -260,9 +260,9 @@ pipeline {
                 sh """
                     export TBB_INC=\$(pwd)/lib/tbb_2020.3/include
                     export TBB_LIB=\$(pwd)/lib/tbb
+		    echo TBB_INTERFACE_NEW=true > make/local
                 """
-	            sh "echo CXX=${CLANG_CXX} >> make/local"
-                sh "echo CXXFLAGS += -fsanitize=address >> make/local"
+	        sh "echo CXXFLAGS += -fsanitize=address >> make/local"
                 script {
                     if (isUnix()) {
                         runTests("test/unit/math/test_ad_test.cpp", false)

--- a/Jenkinsfile
+++ b/Jenkinsfile
@@ -260,7 +260,6 @@ pipeline {
                 sh """
                     export TBB_INC=\$(pwd)/lib/tbb_2020.3/include
                     export TBB_LIB=\$(pwd)/lib/tbb
-                    echo TBB_INTERFACE_NEW=true > make/local
                 """
 	            sh "echo CXX=${CLANG_CXX} >> make/local"
                 sh "echo CXXFLAGS += -fsanitize=address >> make/local"

--- a/Jenkinsfile
+++ b/Jenkinsfile
@@ -260,9 +260,9 @@ pipeline {
 	            sh "echo CXXFLAGS += -fsanitize=address >> make/local"
                 script {
                     if (isUnix()) {
-                        runTests("test/unit/math/test_ad_test.cpp", false)
+                        runTests("test/unit", false)
                     } else {
-                        runTestsWin("test/unit/math/test_ad_test.cpp", true)
+                        runTestsWin("test/unit", true)
                     }
                 }
             }

--- a/Jenkinsfile
+++ b/Jenkinsfile
@@ -243,11 +243,7 @@ pipeline {
         }
 
         stage('Full Unit Tests') {
-            agent {
-                docker {
-                    image 'stanorg/ci:gpu'
-                    label 'linux'
-                }
+            agent {  label 'gg-linux'  }
             }
             when {
                 expression {

--- a/Jenkinsfile
+++ b/Jenkinsfile
@@ -262,7 +262,8 @@ pipeline {
                     export TBB_LIB=\$(pwd)/lib/tbb
                     echo TBB_INTERFACE_NEW=true > make/local
                 """
-	            sh "echo CXXFLAGS += -fsanitize=address >> make/local"
+	            sh "echo CXX=${CLANG_CXX} >> make/local"
+                sh "echo CXXFLAGS += -fsanitize=address >> make/local"
                 script {
                     if (isUnix()) {
                         runTests("test/unit/math/test_ad_test.cpp", false)

--- a/Jenkinsfile
+++ b/Jenkinsfile
@@ -244,7 +244,6 @@ pipeline {
 
         stage('Full Unit Tests') {
             agent {  label 'gg-linux'  }
-            }
             when {
                 expression {
                     !skipRemainingStages
@@ -253,12 +252,12 @@ pipeline {
             steps {
                 unstash 'MathSetup'
                 // Set Stan local compiler flags to use the new TBB interface
-                sh """
-                    export TBB_INC=\$(pwd)/lib/tbb_2020.3/include
-                    export TBB_LIB=\$(pwd)/lib/tbb
-		    echo TBB_INTERFACE_NEW=true > make/local
-                """
-	        sh "echo CXXFLAGS += -fsanitize=address >> make/local"
+                // sh """
+                //     export TBB_INC=\$(pwd)/lib/tbb_2020.3/include
+                //     export TBB_LIB=\$(pwd)/lib/tbb
+                //     echo TBB_INTERFACE_NEW=true > make/local
+                // """
+	            sh "echo CXXFLAGS += -fsanitize=address >> make/local"
                 script {
                     if (isUnix()) {
                         runTests("test/unit/math/test_ad_test.cpp", false)

--- a/stan/math/prim/fun/to_matrix.hpp
+++ b/stan/math/prim/fun/to_matrix.hpp
@@ -34,10 +34,11 @@ inline EigMat to_matrix(EigMat&& x) {
  * @return the matrix representation of the input
  */
 template <typename EigMat, require_eigen_vector_t<EigMat>* = nullptr>
-inline Eigen::Matrix<value_type_t<EigMat>, Eigen::Dynamic, Eigen::Dynamic> to_matrix(
-    const EigMat& matrix) {
+inline Eigen::Matrix<value_type_t<EigMat>, Eigen::Dynamic, Eigen::Dynamic> 
+  to_matrix(const EigMat& matrix) {
   using T = value_type_t<EigMat>;
-  Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic> res(matrix.rows(), matrix.cols());
+  Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic> 
+      res(matrix.rows(), matrix.cols());
   Eigen::Map<
       Eigen::Matrix<T, EigMat::RowsAtCompileTime, EigMat::ColsAtCompileTime>>
       res_map(res.data(), matrix.rows(), matrix.cols());

--- a/stan/math/prim/fun/to_matrix.hpp
+++ b/stan/math/prim/fun/to_matrix.hpp
@@ -35,7 +35,9 @@ inline EigMat to_matrix(EigMat&& x) {
  */
 template <typename EigVec, require_eigen_vector_t<EigVec>* = nullptr>
 inline auto to_matrix(EigVec&& matrix) {
-  return Eigen::Matrix<value_type_t<EigVec>, Eigen::Dynamic, Eigen::Dynamic>(std::forward<EigVec>(matrix));
+  return Eigen::Matrix<
+          value_type_t<EigVec>,
+          Eigen::Dynamic, Eigen::Dynamic>(std::forward<EigVec>(matrix));
 }
 
 /**

--- a/stan/math/prim/fun/to_matrix.hpp
+++ b/stan/math/prim/fun/to_matrix.hpp
@@ -18,7 +18,7 @@ namespace math {
  * @param x matrix
  * @return the matrix representation of the input
  */
-template <typename EigMat, require_eigen_matrix_dynamic_t<EigMat>* = nullptr>
+template <typename EigMat, require_eigen_dense_dynamic_t<EigMat>* = nullptr>
 inline EigMat to_matrix(EigMat&& x) {
   return std::forward<EigMat>(x);
 }
@@ -33,17 +33,9 @@ inline EigMat to_matrix(EigMat&& x) {
  * @param x input vector/row vector
  * @return the matrix representation of the input
  */
-template <typename EigMat, require_eigen_vector_t<EigMat>* = nullptr>
-inline Eigen::Matrix<value_type_t<EigMat>, Eigen::Dynamic, Eigen::Dynamic>
-  to_matrix(const EigMat& matrix) {
-  using T = value_type_t<EigMat>;
-  Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic>
-      res(matrix.rows(), matrix.cols());
-  Eigen::Map<
-      Eigen::Matrix<T, EigMat::RowsAtCompileTime, EigMat::ColsAtCompileTime>>
-      res_map(res.data(), matrix.rows(), matrix.cols());
-  res_map = matrix;
-  return res;
+template <typename EigVec, require_eigen_vector_t<EigVec>* = nullptr>
+inline auto to_matrix(EigVec&& matrix) {
+  return Eigen::Matrix<value_type_t<EigMat>, Eigen::Dynamic, Eigen::Dynamic>(std::forward<EigVec>(matrix));
 }
 
 /**

--- a/stan/math/prim/fun/to_matrix.hpp
+++ b/stan/math/prim/fun/to_matrix.hpp
@@ -11,18 +11,38 @@ namespace math {
 
 /**
  * Returns a matrix with dynamic dimensions constructed from an
- * Eigen matrix which is either a row vector, column vector,
- * or matrix.
- * The runtime dimensions will be the same as the input.
+ * Eigen matrix.
  *
  * @tparam EigMat type of the matrix
  *
  * @param x matrix
  * @return the matrix representation of the input
  */
-template <typename EigMat, require_eigen_t<EigMat>* = nullptr>
+template <typename EigMat, require_eigen_matrix_dynamic_t<EigMat>* = nullptr>
 inline EigMat to_matrix(EigMat&& x) {
   return std::forward<EigMat>(x);
+}
+
+/**
+ * Returns a matrix with dynamic dimensions constructed from an
+ * Eigen row or column vector.
+ * The runtime dimensions will be the same as the input.
+ *
+ * @tparam EigMat type of the vector/row vector
+ *
+ * @param x input vector/row vector
+ * @return the matrix representation of the input
+ */
+template <typename EigMat, require_eigen_vector_t<EigMat>* = nullptr>
+inline Eigen::Matrix<value_type_t<EigMat>, Eigen::Dynamic, Eigen::Dynamic> to_matrix(
+    const EigMat& matrix) {
+  using T = value_type_t<EigMat>;
+  Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic> res(matrix.rows(), matrix.cols());
+  Eigen::Map<
+      Eigen::Matrix<T, EigMat::RowsAtCompileTime, EigMat::ColsAtCompileTime>>
+      res_map(res.data(), matrix.rows(), matrix.cols());
+  res_map = matrix;
+  return res;
 }
 
 /**

--- a/stan/math/prim/fun/to_matrix.hpp
+++ b/stan/math/prim/fun/to_matrix.hpp
@@ -35,7 +35,7 @@ inline EigMat to_matrix(EigMat&& x) {
  */
 template <typename EigVec, require_eigen_vector_t<EigVec>* = nullptr>
 inline auto to_matrix(EigVec&& matrix) {
-  return Eigen::Matrix<value_type_t<EigMat>, Eigen::Dynamic, Eigen::Dynamic>(std::forward<EigVec>(matrix));
+  return Eigen::Matrix<value_type_t<EigVec>, Eigen::Dynamic, Eigen::Dynamic>(std::forward<EigVec>(matrix));
 }
 
 /**

--- a/stan/math/prim/fun/to_matrix.hpp
+++ b/stan/math/prim/fun/to_matrix.hpp
@@ -34,10 +34,10 @@ inline EigMat to_matrix(EigMat&& x) {
  * @return the matrix representation of the input
  */
 template <typename EigMat, require_eigen_vector_t<EigMat>* = nullptr>
-inline Eigen::Matrix<value_type_t<EigMat>, Eigen::Dynamic, Eigen::Dynamic> 
+inline Eigen::Matrix<value_type_t<EigMat>, Eigen::Dynamic, Eigen::Dynamic>
   to_matrix(const EigMat& matrix) {
   using T = value_type_t<EigMat>;
-  Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic> 
+  Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic>
       res(matrix.rows(), matrix.cols());
   Eigen::Map<
       Eigen::Matrix<T, EigMat::RowsAtCompileTime, EigMat::ColsAtCompileTime>>

